.upstream-tests/test/host_only/CMakeLists.txt

```diff
@@ -9,22 +18,46 @@ function(libcudacxx_add_standalone_header_test test_file_output_var header_under
 
     set(${test_file_output_var} ${test_cpp} PARENT_SCOPE)
     # ex: cuda/std/version -> cuda_std_version
-    string(REPLACE "/" "_" executable_name ${header_under_test})
-    add_executable(${executable_name} ${test_cpp})
-    target_include_directories(${executable_name} PRIVATE ${CMAKE_SOURCE_DIR}/include)
-    target_compile_options(${executable_name} PRIVATE
+
+    string(REPLACE "/" "_" executable_name "${header_under_test}-${version}")
+    list(FIND cpp_${version}_exclusions ${header_under_test} header_excluded)
+
+    if ("${header_excluded}" EQUAL "-1")
+        add_library(${executable_name} OBJECT ${test_cpp})
+        target_include_directories(${executable_name} PRIVATE ${CMAKE_SOURCE_DIR}/include)
+        target_compile_options(${executable_name} PRIVATE
         $<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>,$<CXX_COMPILER_ID:GNU>>:
-            -Wall -Werror>
+        -Wall -Werror>
         $<$<CXX_COMPILER_ID:MSVC>:
-            /W4 /WX>)
-    set_target_properties(${executable_name} PROPERTIES CXX_STANDARD 11)
+        /W4 /WX>)
+        set_target_properties(${executable_name} PROPERTIES CXX_STANDARD ${version})
+        set_target_properties(${executable_name} PROPERTIES CXX_STANDARD_REQUIRED ON)
+        set_target_properties(${executable_name} PROPERTIES CXX_EXTENSIONS OFF)
+        add_dependencies(libcudacxx_tu_tests ${executable_name})
+    endif()
+
 endfunction()
 
 # Don't generate CUDA targets, they fail currently.
 # file(GLOB cuda_headers LIST_DIRECTORIES false RELATIVE ${CMAKE_SOURCE_DIR}/include ${CMAKE_SOURCE_DIR}/include/cuda/*)
 file(GLOB cuda_std_headers LIST_DIRECTORIES false RELATIVE ${CMAKE_SOURCE_DIR}/include ${CMAKE_SOURCE_DIR}/include/cuda/std/*)
 
-foreach(header IN LISTS cuda_headers cuda_std_headers)
-    libcudacxx_add_standalone_header_test(test_file ${header})
-    message(STATUS "Detected ${header} ... Writing ${test_file}")
+foreach(version IN LISTS cpp_std_versions)
+    if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+        set(cxx_version "/std:c++${version}")
+    else()
+        set(cxx_version "-std=c++${version}")
+    endif()
+
+    try_compile(
+        dialect_supported
+        ${CMAKE_BINARY_DIR}/dialect_check_${version} "${CMAKE_CURRENT_SOURCE_DIR}/detect_dialect.cpp"
+        COMPILE_DEFINITIONS ${cxx_version}
```

Comment:
I did try to use a CMake generator here switching on the compiler ID for MSVC, but it didn't seem to work in the subprocess. :(

Reply:
What is your end goal here? Is it to just verify that the given compiler supports a given standard level?

Reply:
The `CMAKE_CXX_COMPILE_FEATURES` variable will hold all the known language levels that the compiler supports.
So you can check if `cxx_std_${cxx_version}` is in that list.

Reply:
Would a dialect be included in that list if the compiler only has a subset of available features?

For example, GCC-6 supports 17, but not deduction guides and thus fails to compile some tests when CMake tries to use `c++1z`. There are also some things that fail on the compilers' side and are outside of our control.

Reply:
Yes it contains all dialects for a compiler which have a mode switch ( such as gcc-6 `c++1z` ).

I think it is better to contain a known list of bad configurations and exclude those compared to trying to build a test case that will smoke out all 'bad' compilers.

Reply:
This is something I wanted to use with NVCC as well which maintains its own compiler support matrix. 

I'm not opposed to the changes I think you'd like to see, but a matrix of compiler, c++ dialect, and header files might be beyond my CMake-fu.

Can we make that a future PR?

Reply:
Sure a future PR works for me